### ISSUE TITLE
MIM crashed when pushing block

### ISSUE THREAD

**Issue by deadjdona** [esl/MongooseIM#1569]
MongooseIM version: 2.1
Installed from: src
Erlang/OTP version: 20

Describe the issue.

Not sure what this means, but when when you try to block user in muclight chats, server crashes.

2017-11-14 16:30:18 =ERROR REPORT====
** State machine <0.1662.0> terminating
** Last event in was {xmlstreamelement,{xmlel,<<"iq">>,[{<<"type">>,<<"error">>},{<<"to">>,<<"82fddd03fdc1c67d9e50fcf37be3acbc@xmpp.domain.com">>},{<<"id">>,<<"push">>}],[{xmlel,<<"block">>,[{<<"xmlns">>,<<"urn:xmpp:blocking">>}],[{xmlel,<<"item">>,[{<<"jid">>,<<"bd541e1b536fb3da8b1753f51234f7c79@xmpp.domain.com">>}],[]}]},{xmlel,<<"error">>,[{<<"type">>,<<"cancel">>},{<<"code">>,<<"501">>}],[{xmlel,<<"feature-not-implemented">>,[{<<"xmlns">>,<<"urn:ietf:params:xml:ns:xmpp-stanzas">>}],[]}]}]}}

2017-11-14 16:30:18 =CRASH REPORT====
  crasher:
    initial call: ejabberd_c2s:init/1
    pid: <0.1662.0>
    registered_name: []
    exception exit: {{{badmatch,reply},[{ejabberd_c2s,process_privacy_iq,3,[{file,"/home/ubuntu/MongooseIM/_build/prod/lib/ejabberd/src/ejabberd_c2s.erl"},{line,2355}]},{ejabberd_c2s,process_outgoing_stanza,4,[{file,"/home/ubuntu/MongooseIM/_build/prod/lib/ejabberd/src/ejabberd_c2s.erl"},{line,1028}]},{ejabberd_c2s,process_outgoing_stanza,2,[{file,"/home/ubuntu/MongooseIM/_build/prod/lib/ejabberd/src/ejabberd_c2s.erl"},{line,987}]},{p1_fsm_old,handle_msg,10,[{file,"/home/ubuntu/MongooseIM/_build/prod/lib/ejabberd/src/p1_fsm_old.erl"},{line,576}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,247}]}]},[{p1_fsm_old,terminate,7,[{file,"/home/ubuntu/MongooseIM/_build/prod/lib/ejabberd/src/p1_fsm_old.erl"},{line,729}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,247}]}]}
    ancestors: [ejabberd_c2s_sup,ejabberd_sup,<0.300.0>]
    message_queue_len: 1
    messages: [{'$gen_event',closed}]
    links: [<0.528.0>]
    dictionary: [{'$internal_queue_len',0},{rand_seed,{#{bits => 58,jump => #Fun<rand.8.15449617>,next => #Fun<rand.5.15449617>,type => exrop,uniform => #Fun<rand.6.15449617>,uniform_n => #Fun<rand.7.15449617>,weak_low_bits => 1},[33767379066818971|26066331653449585]}}]
    trap_exit: false
    status: running
    heap_size: 6772
    stack_size: 27
    reductions: 7343881
  neighbours:
2017-11-14 16:30:18 =SUPERVISOR REPORT====
     Supervisor: {local,ejabberd_c2s_sup}
     Context:    child_terminated
     Reason:     {{badmatch,reply},[{ejabberd_c2s,process_privacy_iq,3,[{file,"/home/ubuntu/MongooseIM/_build/prod/lib/ejabberd/src/ejabberd_c2s.erl"},{line,2355}]},{ejabberd_c2s,process_outgoing_stanza,4,[{file,"/home/ubuntu/MongooseIM/_build/prod/lib/ejabberd/src/ejabberd_c2s.erl"},{line,1028}]},{ejabberd_c2s,process_outgoing_stanza,2,[{file,"/home/ubuntu/MongooseIM/_build/prod/lib/ejabberd/src/ejabberd_c2s.erl"},{line,987}]},{p1_fsm_old,handle_msg,10,[{file,"/home/ubuntu/MongooseIM/_build/prod/lib/ejabberd/src/p1_fsm_old.erl"},{line,576}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,247}]}]}
     Offender:   [{pid,<0.1662.0>},{id,undefined},{mfargs,{ejabberd_c2s,start_link,undefined}},{restart_type,temporary},{shutdown,brutal_kill},{child_type,worker}]

**Comment by izigibran** [comment]
MongooseIM version: master
Installed from: src
Erlang/OTP version: 20

Describe the issue:
I have a similar issue here using https://xmpp.org/extensions/xep-0191.html#block

```
2017-11-15 16:55:46.651 [debug] <0.1206.0>@ejabberd_receiver:process_data:317 Received XML on stream = "<iq type='result' id='blocklist1'>
  <blocklist xmlns='urn:xmpp:blocking'>
    <item jid='user2@localhost'/>
  </blocklist>
</iq>"
2017-11-15 16:55:46.651 [debug] <0.1206.0>@shaper:update:60 Tokens: 871 (+17769,-129), delay: 0 ms
2017-11-15 16:55:46.651 [info] <0.1207.0>@ejabberd_c2s:terminate:1582 ({socket_state,fast_tls,{tlssock,#Port<0.14829>,#Port<0.14830>},<0.1206.0>}) Close session for user1@localhost/localhost
2017-11-15 16:55:46.651 [debug] <0.1207.0>@mod_pubsub:check_plugin_features_and_acc_affs:4360 error="cant_purge_items_on_offline" plugin="<<"push">>" user="user1@localhost/localhost"
2017-11-15 16:55:46.651 [debug] <0.1207.0>@mod_carboncopy:disable:273 disabling for user1@localhost/localhost
2017-11-15 16:55:46.653 [debug] <0.1207.0>@ejabberd_router:route:106 route
        from {jid,<<"user1">>,<<"localhost">>,<<"localhost">>,<<"user1">>,<<"localhost">>,<<"localhost">>}
        to {jid,<<"user1">>,<<"localhost">>,<<>>,<<"user1">>,<<"localhost">>,<<>>}
        packet #{attrs => [{<<"type">>,<<"unavailable">>}],element => {xmlel,<<"presence">>,[{<<"type">>,<<"unavailable">>}],[]},from => <<"user1@localhost/localhost">>,from_jid => {jid,<<"user1">>,<<"localhost">>,<<"localhost">>,<<"user1">>,<<"localhost">>,<<"localhost">>},mongoose_acc => true,name => <<"presence">>,ref => #Ref<0.2358130774.909901829.129944>,result => allow,timestamp => {1510,782946,651441},type => <<"unavailable">>,{cached_privacy_check,<<"localhost">>,<<"user1">>,<<"user1@localhost/localhost">>,<<"user1@localhost">>,<<"presence">>,<<"unavailable">>,out} => allow}
2017-11-15 16:55:46.653 [debug] <0.1207.0>@ejabberd_router:route:450 Using module mongoose_router_global
2017-11-15 16:55:46.653 [debug] <0.1207.0>@ejabberd_router:route:464 filter passed
2017-11-15 16:55:46.653 [debug] <0.1207.0>@ejabberd_router:route:480 routing skipped
2017-11-15 16:55:46.653 [debug] <0.1207.0>@ejabberd_router:route:450 Using module mongoose_router_localdomain
2017-11-15 16:55:46.653 [debug] <0.1207.0>@ejabberd_router:route:464 filter passed
2017-11-15 16:55:46.653 [debug] <0.1207.0>@mod_mam:filter_packet:314 Receive packet
    from {jid,<<"user1">>,<<"localhost">>,<<"localhost">>,<<"user1">>,<<"localhost">>,<<"localhost">>} 
    to {jid,<<"user1">>,<<"localhost">>,<<>>,<<"user1">>,<<"localhost">>,<<>>}
    packet {xmlel,<<"presence">>,[{<<"type">>,<<"unavailable">>}],[]}.
2017-11-15 16:55:46.653 [debug] <0.1207.0>@mod_push:filter_packet:117 Receive packet
    from {jid,<<"user1">>,<<"localhost">>,<<"localhost">>,<<"user1">>,<<"localhost">>,<<"localhost">>} 
    to {jid,<<"user1">>,<<"localhost">>,<<>>,<<"user1">>,<<"localhost">>,<<>>}
    packet #{attrs => [{<<"type">>,<<"unavailable">>}],element => {xmlel,<<"presence">>,[{<<"type">>,<<"unavailable">>}],[]},from => <<"user1@localhost/localhost">>,from_jid => {jid,<<"user1">>,<<"localhost">>,<<"localhost">>,<<"user1">>,<<"localhost">>,<<"localhost">>},mongoose_acc => true,name => <<"presence">>,ref => #Ref<0.2358130774.909901829.129944>,result => allow,timestamp => {1510,782946,651441},type => <<"unavailable">>,{cached_privacy_check,<<"localhost">>,<<"user1">>,<<"user1@localhost/localhost">>,<<"user1@localhost">>,<<"presence">>,<<"unavailable">>,out} => allow}.
2017-11-15 16:55:46.653 [debug] <0.1207.0>@ejabberd_local:do_route:392 local route
        from {jid,<<"user1">>,<<"localhost">>,<<"localhost">>,<<"user1">>,<<"localhost">>,<<"localhost">>}
        to {jid,<<"user1">>,<<"localhost">>,<<>>,<<"user1">>,<<"localhost">>,<<>>}
        packet {xmlel,<<"presence">>,[{<<"type">>,<<"unav"...>>}],[]}
2017-11-15 16:55:46.653 [debug] <0.1207.0>@ejabberd_sm:do_route:608 session manager
        from {jid,<<"user1">>,<<"localhost">>,<<"localhost">>,<<"user1">>,<<"localhost">>,<<"localhost">>}
        to {jid,<<"user1">>,<<"localhost">>,<<>>,<<"user1">>,<<"localhost">>,<<>>}
        packet #{attrs => [{<<"type">>,<<"unavailable">>}],element => {xmlel,<<"presence">>,[{<<...>>,...}],[]},from => <<"user1@localhost/"...>>,from_jid => {jid,<<"user"...>>,<<...>>,...},mongoose_acc => true,name => <<"pres"...>>,ref => #Ref<0.2358130774.909901829.129944>,...}
2017-11-15 16:55:46.653 [debug] <0.1207.0>@ejabberd_router:route:476 routing done
2017-11-15 16:55:46.654 [debug] <0.1207.0>@ejabberd_router:route:106 route
        from {jid,<<"user1">>,<<"localhost">>,<<"localhost">>,<<"user1">>,<<"localhost">>,<<"localhost">>}
        to {jid,<<"user2">>,<<"localhost">>,<<>>,<<"user2">>,<<"localhost">>,<<>>}
        packet #{attrs => [{<<"type">>,<<"unavailable">>}],element => {xmlel,<<"presence">>,[{<<"type">>,<<"unavailable">>}],[]},from => <<"user1@localhost/localhost">>,from_jid => {jid,<<"user1">>,<<"localhost">>,<<"localhost">>,<<"user1">>,<<"localhost">>,<<"localhost">>},mongoose_acc => true,name => <<"presence">>,ref => #Ref<0.2358130774.909901829.129944>,result => allow,send_result => [done],timestamp => {1510,782946,651441},type => <<"unavailable">>,{cached_privacy_check,<<"localhost">>,<<"user1">>,<<"user1@localhost/localhost">>,<<"user1@localhost">>,<<"presence">>,<<"unavailable">>,out} => allow,{cached_privacy_check,<<"localhost">>,<<"user1">>,<<"user1@localhost/localhost">>,<<"user2@localhost">>,<<"presence">>,<<"unavailable">>,out} => allow}
2017-11-15 16:55:46.654 [debug] <0.1207.0>@ejabberd_router:route:450 Using module mongoose_router_global
2017-11-15 16:55:46.654 [debug] <0.1207.0>@ejabberd_router:route:464 filter passed
2017-11-15 16:55:46.654 [debug] <0.1207.0>@ejabberd_router:route:480 routing skipped
2017-11-15 16:55:46.654 [debug] <0.1207.0>@ejabberd_router:route:450 Using module mongoose_router_localdomain
2017-11-15 16:55:46.654 [debug] <0.1207.0>@ejabberd_router:route:464 filter passed
2017-11-15 16:55:46.654 [debug] <0.1207.0>@mod_mam:filter_packet:314 Receive packet
    from {jid,<<"user1">>,<<"localhost">>,<<"localhost">>,<<"user1">>,<<"localhost">>,<<"localhost">>} 
    to {jid,<<"user2">>,<<"localhost">>,<<>>,<<"user2">>,<<"localhost">>,<<>>}
    packet {xmlel,<<"presence">>,[{<<"type">>,<<"unavailable">>}],[]}.
2017-11-15 16:55:46.654 [debug] <0.1207.0>@mod_push:filter_packet:117 Receive packet
    from {jid,<<"user1">>,<<"localhost">>,<<"localhost">>,<<"user1">>,<<"localhost">>,<<"localhost">>} 
    to {jid,<<"user2">>,<<"localhost">>,<<>>,<<"user2">>,<<"localhost">>,<<>>}
    packet #{attrs => [{<<"type">>,<<"unavailable">>}],element => {xmlel,<<"presence">>,[{<<"type">>,<<"unavailable">>}],[]},from => <<"user1@localhost/localhost">>,from_jid => {jid,<<"user1">>,<<"localhost">>,<<"localhost">>,<<"user1">>,<<"localhost">>,<<"localhost">>},mongoose_acc => true,name => <<"presence">>,ref => #Ref<0.2358130774.909901829.129944>,result => allow,send_result => [done],timestamp => {1510,782946,651441},type => <<"unavailable">>,{cached_privacy_check,<<"localhost">>,<<"user1">>,<<"user1@localhost/localhost">>,<<"user1@localhost">>,<<"presence">>,<<"unavailable">>,out} => allow,{cached_privacy_check,<<"localhost">>,<<"user1">>,<<"user1@localhost/localhost">>,<<"user2@localhost">>,<<"presence">>,<<"unavailable">>,out} => allow}.
2017-11-15 16:55:46.654 [debug] <0.1207.0>@ejabberd_local:do_route:392 local route
        from {jid,<<"user1">>,<<"localhost">>,<<"localhost">>,<<"user1">>,<<"localhost">>,<<"localhost">>}
        to {jid,<<"user2">>,<<"localhost">>,<<>>,<<"user2">>,<<"localhost">>,<<>>}
        packet {xmlel,<<"presence">>,[{<<"type">>,<<"unav"...>>}],[]}
2017-11-15 16:55:46.654 [debug] <0.1207.0>@ejabberd_sm:do_route:608 session manager
        from {jid,<<"user1">>,<<"localhost">>,<<"localhost">>,<<"user1">>,<<"localhost">>,<<"localhost">>}
        to {jid,<<"user2">>,<<"localhost">>,<<>>,<<"user2">>,<<"localhost">>,<<>>}
        packet #{attrs => [{<<"type">>,<<"unavailable">>}],element => {xmlel,<<"presence">>,[{<<...>>,...}],[]},from => <<"user1@localhost/"...>>,from_jid => {jid,<<"user"...>>,<<...>>,...},mongoose_acc => true,name => <<"pres"...>>,ref => #Ref<0.2358130774.909901829.129944>,...}
2017-11-15 16:55:46.654 [debug] <0.1207.0>@ejabberd_router:route:476 routing done
2017-11-15 16:55:46.655 [debug] <0.1207.0>@ejabberd_router:route:106 route
        from {jid,<<"user1">>,<<"localhost">>,<<"localhost">>,<<"user1">>,<<"localhost">>,<<"localhost">>}
        to {jid,<<"user3">>,<<"localhost">>,<<>>,<<"user3">>,<<"localhost">>,<<>>}
        packet #{attrs => [{<<"type">>,<<"unavailable">>}],element => {xmlel,<<"presence">>,[{<<"type">>,<<"unavailable">>}],[]},from => <<"user1@localhost/localhost">>,from_jid => {jid,<<"user1">>,<<"localhost">>,<<"localhost">>,<<"user1">>,<<"localhost">>,<<"localhost">>},mongoose_acc => true,name => <<"presence">>,ref => #Ref<0.2358130774.909901829.129944>,result => allow,send_result => [done,done],timestamp => {1510,782946,651441},type => <<"unavailable">>,{cached_privacy_check,<<"localhost">>,<<"user1">>,<<"user1@localhost/localhost">>,<<"user1@localhost">>,<<"presence">>,<<"unavailable">>,out} => allow,{cached_privacy_check,<<"localhost">>,<<"user1">>,<<"user1@localhost/localhost">>,<<"user2@localhost">>,<<"presence">>,<<"unavailable">>,out} => allow,{cached_privacy_check,<<"localhost">>,<<"user1">>,<<"user1@localhost/localhost">>,<<"user3@localhost">>,<<"presence">>,<<"unavailable">>,out} => allow}
2017-11-15 16:55:46.655 [debug] <0.1207.0>@ejabberd_router:route:450 Using module mongoose_router_global
2017-11-15 16:55:46.655 [debug] <0.1207.0>@ejabberd_router:route:464 filter passed
2017-11-15 16:55:46.655 [debug] <0.1207.0>@ejabberd_router:route:480 routing skipped
2017-11-15 16:55:46.655 [debug] <0.1207.0>@ejabberd_router:route:450 Using module mongoose_router_localdomain
2017-11-15 16:55:46.655 [debug] <0.1207.0>@ejabberd_router:route:464 filter passed
2017-11-15 16:55:46.655 [debug] <0.1207.0>@mod_mam:filter_packet:314 Receive packet
    from {jid,<<"user1">>,<<"localhost">>,<<"localhost">>,<<"user1">>,<<"localhost">>,<<"localhost">>} 
    to {jid,<<"user3">>,<<"localhost">>,<<>>,<<"user3">>,<<"localhost">>,<<>>}
    packet {xmlel,<<"presence">>,[{<<"type">>,<<"unavailable">>}],[]}.
2017-11-15 16:55:46.655 [debug] <0.1207.0>@mod_push:filter_packet:117 Receive packet
    from {jid,<<"user1">>,<<"localhost">>,<<"localhost">>,<<"user1">>,<<"localhost">>,<<"localhost">>} 
    to {jid,<<"user3">>,<<"localhost">>,<<>>,<<"user3">>,<<"localhost">>,<<>>}
    packet #{attrs => [{<<"type">>,<<"unavailable">>}],element => {xmlel,<<"presence">>,[{<<"type">>,<<"unavailable">>}],[]},from => <<"user1@localhost/localhost">>,from_jid => {jid,<<"user1">>,<<"localhost">>,<<"localhost">>,<<"user1">>,<<"localhost">>,<<"localhost">>},mongoose_acc => true,name => <<"presence">>,ref => #Ref<0.2358130774.909901829.129944>,result => allow,send_result => [done,done],timestamp => {1510,782946,651441},type => <<"unavailable">>,{cached_privacy_check,<<"localhost">>,<<"user1">>,<<"user1@localhost/localhost">>,<<"user1@localhost">>,<<"presence">>,<<"unavailable">>,out} => allow,{cached_privacy_check,<<"localhost">>,<<"user1">>,<<"user1@localhost/localhost">>,<<"user2@localhost">>,<<"presence">>,<<"unavailable">>,out} => allow,{cached_privacy_check,<<"localhost">>,<<"user1">>,<<"user1@localhost/localhost">>,<<"user3@localhost">>,<<"presence">>,<<"unavailable">>,out} => allow}.
2017-11-15 16:55:46.655 [debug] <0.1207.0>@ejabberd_local:do_route:392 local route
        from {jid,<<"user1">>,<<"localhost">>,<<"localhost">>,<<"user1">>,<<"localhost">>,<<"localhost">>}
        to {jid,<<"user3">>,<<"localhost">>,<<>>,<<"user3">>,<<"localhost">>,<<>>}
        packet {xmlel,<<"presence">>,[{<<"type">>,<<"unav"...>>}],[]}
2017-11-15 16:55:46.655 [debug] <0.1207.0>@ejabberd_sm:do_route:608 session manager
        from {jid,<<"user1">>,<<"localhost">>,<<"localhost">>,<<"user1">>,<<"localhost">>,<<"localhost">>}
        to {jid,<<"user3">>,<<"localhost">>,<<>>,<<"user3">>,<<"localhost">>,<<>>}
        packet #{attrs => [{<<"type">>,<<"unavailable">>}],element => {xmlel,<<"presence">>,[{<<...>>,...}],[]},from => <<"user1@localhost/"...>>,from_jid => {jid,<<"user"...>>,<<...>>,...},mongoose_acc => true,name => <<"pres"...>>,ref => #Ref<0.2358130774.909901829.129944>,...}
2017-11-15 16:55:46.655 [debug] <0.1207.0>@ejabberd_router:route:476 routing done
2017-11-15 16:55:46.655 [debug] <0.1207.0>@ejabberd_c2s:reroute_unacked_messages:1620 rerouting unacked messages
2017-11-15 16:55:46.655 [error] <0.1207.0> gen_fsm <0.1207.0> in state session_established terminated with reason: no match of right hand value reply in ejabberd_c2s:process_privacy_iq/3 line 2355
2017-11-15 16:55:46.656 [error] <0.1207.0> CRASH REPORT Process <0.1207.0> with 0 neighbours exited with reason: no match of right hand value reply in ejabberd_c2s:process_privacy_iq/3 line 2355 in p1_fsm_old:terminate/7 line 729
2017-11-15 16:55:46.656 [error] <0.433.0> Supervisor ejabberd_c2s_sup had child undefined started with {ejabberd_c2s,start_link,undefined} at <0.1207.0> exit with reason no match of right hand value reply in ejabberd_c2s:process_privacy_iq/3 line 2355 in context child_terminated
```

**Comment by fenek** [comment]
Hi @deadjdona 

Indeed it is a bug in the server. It should ignore such an error from a client instead of crashing. Thanks for spotting this!

Anyway, are you using some desktop client? If so, then this problem should be observable in actual custom application, because the client responded with "feature not implemented" for server's push with blocklist update. In real world it doesn't make sense for the client to implement Blocking Command request without handling server push properly.

**Comment by deadjdona** [comment]
@fenek I use android and ios clients with Smack and xmppframework

**Comment by fenek** [comment]
Hmm... OK, are you using some built-in Blocking Command extensions for these libraries or are you creating the IQ request on your own?

**Comment by deadjdona** [comment]
built-in functionality.

**Comment by fenek** [comment]
I can't reproduce the issue. The client replies with `result`, not `error`:

```java
BlockingCommandManager blockingCommandManager = BlockingCommandManager.getInstanceFor(connection);
List<Jid> jidList = new LinkedList<Jid>();
jidList.add(JidCreate.from("b", "localhost", ""));
blockingCommandManager.blockContacts(jidList);
```

I tried it with Smack 4.2.2.

**Comment by fenek** [comment]
I've merged #1597, it should fix your issue. :)
